KAFKA-2111: Add help arguments and required fields

johnma14 opened the pull request:
This patch addresses the less invasive standardization of command
line arguments as an offshoot of KIP-14. The following changes have
been addressed in this patchset:
- Used the required() method of JOptSimple to identify required options
  for commands. Currently, an option is identified as required by
  mentioning the word REQUIRED in the description of the option.
- Used the requiredIf() and requiredUnless() methods to capture option
  requirements which met the condition - required if/unless some other
  option is present
- Removed calls to CommandLineUtils.checkRequiredArgs(). Since we use
  JOptSimple's required() function to mark options as required, when
  we parse the arguments using OptionParser, it automatically checks
  the required arguments in that call. So there is no need to do this
  check again externally.
- Added a help option to all the commands that were missing it. For
  some of the commands that had the help option, included the call to
  forHelp() method to mark the option as a help option. This means that
  specifying help option on the command line will not cause parsing to
  fail because of missing required options.
- Added try-catch statements to correctly capture OptionException. In
  the case an OptionException is caught, the command usage information
  will be printed along with the exception message. The stacktrace for
  all other exceptions will be printed to System.err. This addresses
  the issue in KAFKA-4220
- Made some minor changes to the description text and added some more
  description for some of the options

johnma14 committed:
KAFKA-2111: Add help arguments and required fields

This patch addresses the less invasive standardization of command
line arguments as an offshoot of KIP-14. The following changes have
been addressed in this patchset:
- Used the required() method of JOptSimple to identify required options
  for commands. Currently, an option is identified as required by
  mentioning the word REQUIRED in the description of the option.
- Used the requiredIf() and requiredUnless() methods to capture option
  requirements which met the condition - required if/unless some other
  option is present
- Removed calls to CommandLineUtils.checkRequiredArgs(). Since we use
  JOptSimple's required() function to mark options as required, when
  we parse the arguments using OptionParser, it automatically checks
  the required arguments in that call. So there is no need to do this
  check again externally.
- Added a help option to all the commands that were missing it. For
  some of the commands that had the help option, included the call to
  forHelp() method to mark the option as a help option. This means that
  specifying help option on the command line will not cause parsing to
  fail because of missing required options.
- Made some minor changes to the description text and added some more
  description for some of the options

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6453/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6468/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
FAILURE
 No test results found.
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6454/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6469/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6459/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6474/
Test FAILed (JDK 7 and Scala 2.11).

johnma14 commented:
retest please

johnma14 commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6463/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6478/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6665/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6650/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6668/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6653/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6693/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6678/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6720/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6705/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6753/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6738/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6789/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6775/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6799/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6784/
Test PASSed (JDK 8 and Scala 2.12).

johnma14 commented:
@ijuma @guozhangwang @mjsax @dguy This patch is ready for review whenever it is a good time for you. In this patch set I have taken a stab at trying to standardize how we define some of the options for Kafka commands as required, include a help message for these commands and fix the test cases to reflect these changes. Thank you for your feedback and appreciate your time.

vahidhashemian commented on core/src/main/scala/kafka/utils/CommandLineUtils.scala:
```diff
@@ -76,4 +76,16 @@ object CommandLineUtils extends Logging {
+  /**
+   * 
+   */
+  def tryParse(parser: OptionParser, args: Array[String]): OptionSet = {
```
- If there is no Javadoc please remove the comments.

johnma14 replied:
Fixed

vahidhashemian commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -323,30 +344,33 @@ object TopicCommand extends Logging {
+    val forceOpt = parser.accepts("force", "Suppress console prompts.")
 
-    val forceOpt = parser.accepts("force", "Suppress console prompts")
+    val commandDef:String = "Create, delete, describe, or change a topic."
```
Missing a whitespace after the colon: `commandDef: String`

johnma14 replied:
Done

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -271,8 +271,9 @@ object ConsoleConsumer extends Logging {
-      "Multiple URLS can be given to allow fail-over.")
+    val zkConnectOpt = parser.accepts("zookeeper", "The connection string for the zookeeper connection in the form host:port. " +
+      "This is required only when using old consumer (By default this command uses the new consumer implementation. The old consumer implementation " +
+      "is used if the zookeeper option is specified). Multiple URLS can be given to allow fail-over.")
```
Doesn't this need `.requiredUnless("bootstrap-server")`?

johnma14 replied:
we can only specify requiredUnless on either bootstrap-server or zookeeper options and not both. When we specify A.requiredUnless(B), then B needs to be defined first.

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -306,15 +307,17 @@ object ConsoleConsumer extends Logging {
-    val bootstrapServerOpt = parser.accepts("bootstrap-server", "REQUIRED (unless old consumer is used): The server to connect to.")
+    val bootstrapServerOpt = parser.accepts("bootstrap-server", "The server to connect to. This is required with the new consumer implementation." +
+        "(The new consumer implementation is used by default or can be explicitely set using the --new-consumer option.)")
+      .requiredUnless("zookeeper")
```
One minor issue with this new approach is that for a command like this console consumer, where there is no always required option, the user who is new to the tool would have to read through the help doc to find out which ones are required (because no option would get the `*` mark). So far the capital `REQUIRED` in the help doc easily catches the eye. It's not a big deal, as the method used in this PR to handle command line arguments seems to be much cleaner.

johnma14 replied:
I agree. I debated this issue alot in my mind but this seemed to be the consistent way to fix this across all commands using OptionParser.

vahidhashemian commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -142,7 +146,13 @@ object TopicCommand extends Logging {
+          nPartitions = opts.options.valueOf(opts.partitionsOpt).intValue
+        } catch {
+          case ex: OptionException =>
+            CommandLineUtils.printUsageAndDie(opts.parser, ex.getMessage)
```
In my opinion the existing behavior where an error is returned to the user (without the help doc) is more user friendly. The issue with the proposed method is that the error message is printed first and then the normally long help doc, so user has to scroll back up to see what went wrong. This applies to other tools too. Is there a particular reason why we are printing the help doc now?

As I said, this is my opinion. So maybe others should also weigh in on this before changing it.

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -134,17 +135,19 @@ object ConsoleProducer {
       .ofType(classOf[String])
-    val brokerListOpt = parser.accepts("broker-list", "REQUIRED: The broker list string in the form HOST1:PORT1,HOST2:PORT2.")
+      .required
+    val brokerListOpt = parser.accepts("broker-list", "The broker list string in the form HOST1:PORT1,HOST2:PORT2.")
```
For consistency, can we use lowercase `host1:port1, ...` like in other help docs?

johnma14 replied:
done

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -265,25 +276,31 @@ object ConsoleProducer {
+    } catch {
+      case ex: OptionException =>
+        CommandLineUtils.printUsageAndDie(parser, ex.getMessage)
+    }
```
Another issue with this new way of reporting an error and then printing the help doc: If I use an invalid value for `--request-required-acks` (e.g. `abc`), I just get an error. But if I use an invalid value for `--batch-size` (e.g. `abc`) I get an error and the help doc.

This is currently handled better (not perfect though, as one prints an exception too, but the other just prints an error message) in my opinion.

johnma14 replied:
For now I am removing these exception handling out of this pull request.

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -243,57 +248,76 @@ object ConsumerPerformance {
+    
+    if (options.has("help")) {
+       CommandLineUtils.printUsageAndDie(parser, commandDef)
+    }
```
nit: brackets can be dropped for one-liner blocks.

johnma14 replied:
Done

vahidhashemian commented on core/src/main/scala/kafka/tools/ReplayLogProducer.scala:
```diff
@@ -31,7 +31,9 @@ object ReplayLogProducer extends Logging {
+    var config: Config = null
+    try {
+    config = new Config(args)
 
```
Code inside the `try` block is not indented.

vahidhashemian commented on core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala:
```diff
@@ -113,86 +119,89 @@ object ReplicaVerificationTool extends Logging {
+            true
+          else
+            false
+      )
```
I wonder if we can use this instead:
```
      val filteredTopicMetadata = topicsMetadataResponse.topicsMetadata.filter(
        topicMetadata => topicWhiteListFiler.isTopicAllowed(topicMetadata.topic, excludeInternalTopics = false))
```

johnma14 replied:
Done

vahidhashemian reviewed (changes requested):
@johnma14 Thanks for attempting to clean up argument handling in the tools. I left some comments inline.

guozhangwang commented:
@johnma14 Could you rebase?

johnma14 commented:
retest this please

johnma14 commented:
retest this please

johnma14 commented:
retest this please

johnma14 commented:
@guozhangwang @vahidhashemian the tests have passed and this patch is ready for review, whenever you have some time. Appreciate it, thank you.

vahidhashemian commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -278,71 +275,83 @@ object ConfigCommand extends Config {
         if (entityTypeVals.contains(ConfigType.User) || entityTypeVals.contains(ConfigType.Client)) {
           if (!options.has(entityName) && !options.has(entityDefault))
-            throw new IllegalArgumentException("--entity-name or --entity-default must be specified with --alter of users/clients")
+            throw new IllegalArgumentException( "--entity-name or --entity-default must be specified with --alter of users/clients.")
```
nit: The whitespace before the string should be removed.

vahidhashemian commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -278,71 +275,83 @@ object ConfigCommand extends Config {
       )
       if (entityTypeVals.isEmpty)
-        throw new IllegalArgumentException("At least one --entity-type must be specified")
+        throw new IllegalArgumentException( "At least one --entity-type must be specified")
```
nit: The whitespace before the string should be removed. Also a `.` at the end to keep it consistent with the other error messages.

vahidhashemian commented on core/src/main/scala/kafka/tools/JmxTool.scala:
```diff
@@ -124,29 +121,7 @@ object JmxTool extends Logging {
-      System.err.println(s"Could not find all requested object names after $waitTimeoutMs ms. Missing $missing")
-      System.err.println("Exiting.")
-      sys.exit(1)
-    }
```
Is this block of code not needed anymore?

vahidhashemian commented on core/src/main/scala/kafka/tools/SimpleConsumerShell.scala:
```diff
@@ -128,9 +133,9 @@ object SimpleConsumerShell extends Logging {
     // getting topic metadata
-    info("Getting topic metadata...")
     val brokerList = options.valueOf(brokerListOpt)
     ToolsUtils.validatePortOrDie(parser,brokerList)
```
nit: A whitespace after `,` is missing.

vahidhashemian commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -314,32 +312,32 @@ object AclCommand {
       .ofType(classOf[String])
```
It seems in most tools, when `.withRequiredArgs` is used, `.describedAs(...)` is also used to describe the argument required for the option. It would be great if we can use it everywhere to avoid inconsistencies like this (`<String: group>` vs `<String>`):
```
--group <String: group>             Consumer Group to which the ACLs       
                                      should be added or removed. A value  
                                      of * indicates the ACLs should apply 
                                      to all groups.
...
--operation <String>                Operation that is being allowed or     
                                      denied. Valid operation names are:
```

I also think the string used in `describedAs(...)` for some tools parameters is not descriptive enough, but that's a different story :)

johnma14 replied:
Vahid, so I added describedAs to all the options and have also made an attempt at making it more descriptive. I need to scan these files one more time to make sure I haven't missed anything.

vahidhashemian commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -278,71 +275,83 @@ object ConfigCommand extends Config {
+                                  "If --entity-type is clients/users, --entity-name or --entity-default option must be specified. For all other values " +
+                                  "of --entity-type, it is required to specify --entity-name. \nAt least one of options --add-config or --delete-config " +
+                                  "must be specified.")
+    val describeOpt = parser.accepts("describe", "List configurations for the given entity. The entity type must be specified using the --entity-type option.")
```
It seems that either `--alter` or `--describe` is required for any `--entity-type`. If so, does it make sense to use `.requiredUnless` for one of them here?

vahidhashemian commented on core/src/main/scala/kafka/tools/ExportZkOffsets.scala:
```diff
@@ -54,26 +54,25 @@ object ExportZkOffsets extends Logging {
                             .withRequiredArg()
                             .defaultsTo("localhost:2181")
```
If I don't provide the `--zkconnect` option, I get `Missing required option(s) [zkconnect]`. So, `defaultsTo` doesn't seem to work. I think it should either be removed or fixed (if you think this fits under this PR). This seems to work fine for `ImportZkOffsets`. Also, if we get `defaultsTo` working, this option should no longer be required.

vahidhashemian commented on core/src/main/scala/kafka/tools/ProducerPerformance.scala:
```diff
@@ -40,8 +40,10 @@ import org.apache.log4j.Logger
 object ProducerPerformance extends Logging {
```
Not sure if I asked this before. Why is the help doc of this tool different from the other ones?

johnma14 replied:
So the command line uses the class ProducerPerformance.java under tools/src/main/java/org/apache/kafka/tools (which uses ArgumentParser) instead of this scala class(which uses JOptParser). I still just went ahead and made changes here to make it consistent with changes made in other scala classes.

vahidhashemian commented on core/src/main/scala/kafka/tools/ReplayLogProducer.scala:
```diff
@@ -62,24 +63,27 @@ object ReplayLogProducer extends Logging {
       .withRequiredArg
       .describedAs("zookeeper url")
       .ofType(classOf[String])
       .defaultsTo("127.0.0.1:2181")
```
Some commands use `localhost:xxxx` and some use `127.0.0.1:xxxx`. Would be great if we use a single format everywhere.

vahidhashemian reviewed (commented):
@johnma14 Thanks for addressing my earlier comments, and submitting an update. I left some minor comments. LGTM otherwise.

johnma14 commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -278,71 +275,90 @@ object ConfigCommand extends Config {
-    val entityType = parser.accepts("entity-type", "Type of entity (topics/clients/users/brokers)")
+            .required
+    val alterOpt = parser.accepts("alter", "Alter the configuration for the entity. This is required unless the user specifies --describe option. The " +
+                                  "entity type must be specified using option --entity-type. " +
```
Vahid, this is to address your comment about adding requiredUnless for alter and describe options since either one of those action was required. 

RequiredUnless can only be applied to one of the options: alter or describe and not both at the same time. If I apply it to on one of them, say for example we have 2 options : A and B and we say: B is requiredUnless A and in the command line we do not mention A or B, it will print an error message saying B is required. Ideally, we would want it to print 'A or B is required" and it does not look like JOptParser has a way to do this. 

So what I ended up doing is not use requiredUnless but instead in the description for alter and describe options, I mentioned that those options are required unless the other is specified. Hope that helps.

johnma14 commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -278,71 +275,90 @@ object ConfigCommand extends Config {
         if (entityTypeVals.contains(ConfigType.User) || entityTypeVals.contains(ConfigType.Client)) {
           if (!options.has(entityName) && !options.has(entityDefault))
-            throw new IllegalArgumentException("--entity-name or --entity-default must be specified with --alter of users/clients")
+            throw new IllegalArgumentException("--entity-name or --entity-default must be specified with --alter of users/clients.")
```
removed the extra space before the start of this message.

johnma14 commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -278,71 +275,90 @@ object ConfigCommand extends Config {
-          throw new IllegalArgumentException("At least one of --add-config or --delete-config must be specified with --alter")
+        if(!isAddConfigPresent && !isDeleteConfigPresent)
+          throw new IllegalArgumentException("At least one of --add-config or --delete-config must be specified with --alter.")
       }
```
removed white space before the message as per your suggestions.

johnma14 commented on core/src/main/scala/kafka/tools/JmxTool.scala:
```diff
@@ -143,7 +143,7 @@ object JmxTool extends Logging {
       val missing = (queries.toSet - namesSet).mkString(", ")
-      System.err.println(s"Could not find all requested object names after $waitTimeoutMs ms. Missing $missing")
+      System.err.println(s"Could not find all requested object names after $waitTimeoutMs ms. Missing $missing.")
       System.err.println("Exiting.")
```
I added back some of the lines of code that I had unknowingly removed from an earlier commit. Thanks for catching this Vahid.

johnma14 commented on core/src/main/scala/kafka/tools/ExportZkOffsets.scala:
```diff
@@ -52,28 +52,29 @@ object ExportZkOffsets extends Logging {
                             .withRequiredArg()
+                            .describedAs("url(s) for the zookeeper connection")
                             .defaultsTo("localhost:2181")
                             .ofType(classOf[String])
```
https://cwiki.apache.org/confluence/display/KAFKA/System+Tools#SystemTools-ExportZookeeperOffsets
In the above link, it shows that zkconnect is a required option but it also has a defaultsTo value set. So the defaultsTo value does not work if we do not specify zkconnect. Per Vahid's suggestion, I went ahead and made this a regular option(not a required option) with the defaultsTo value and instead made the output file required (this way it is consistent with ImportZkOffsets). Thanks Vahid

johnma14 commented:
retest this please

vahidhashemian commented on core/src/main/scala/kafka/admin/BrokerApiVersionsCommand.scala:
```diff
@@ -70,14 +71,17 @@ object BrokerApiVersionsCommand {
     val bootstrapServerOpt = parser.accepts("bootstrap-server", BootstrapServerDoc)
                                    .withRequiredArg
-                                   .describedAs("server(s) to use for bootstrapping")
+                                   .describedAs("server(s) to connect to")
```
I think the previous description is better, since it emphasizes on bootstrapping. It would be great if we use something similar in other places with a `bootstrap-server` argument.

vahidhashemian commented on core/src/main/scala/kafka/admin/BrokerApiVersionsCommand.scala:
```diff
@@ -70,14 +71,17 @@ object BrokerApiVersionsCommand {
-      CommandLineUtils.checkRequiredArgs(parser, options, bootstrapServerOpt)
-    }
+    var commandDef = "Retrieve broker version information."
+    if(args.length == 0)
```
nit: missing whitespace after `if`

vahidhashemian commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -738,31 +735,33 @@ object ConsumerGroupCommand extends Logging {
     val ResetToCurrentDoc = "Reset offsets to current offset."
     val ResetShiftByDoc = "Reset offsets shifting current offset by 'n', where 'n' can be positive or negative"
 
+
```
nit: extra blank line

vahidhashemian commented on core/src/main/scala/kafka/admin/DeleteRecordsCommand.scala:
```diff
@@ -22,7 +22,7 @@ import java.util.Properties
 import kafka.admin.AdminClient.DeleteRecordsResult
 import kafka.common.AdminCommandFailedException
-import kafka.utils.{CoreUtils, Json, CommandLineUtils}
+import kafka.utils.{CoreUtils, Exit, Json, CommandLineUtils}
```
It seems the import `Exit` is not used.

vahidhashemian commented on core/src/main/scala/kafka/admin/DeleteRecordsCommand.scala:
```diff
@@ -51,7 +51,8 @@ object DeleteRecordsCommand {
   def execute(args: Array[String], out: PrintStream): Unit = {
-    val opts = new DeleteRecordsCommandOptions(args)
+    var opts: DeleteRecordsCommandOptions = null
+    opts = new DeleteRecordsCommandOptions(args)
```
Why was this changed? The old code seems to work.

vahidhashemian commented on core/src/main/scala/kafka/admin/DeleteRecordsCommand.scala:
```diff
@@ -84,26 +84,34 @@ object DeleteRecordsCommand {
     val bootstrapServerOpt = parser.accepts("bootstrap-server", BootstrapServerDoc)
                                    .withRequiredArg
-                                   .describedAs("server(s) to use for bootstrapping")
+                                   .describedAs("server(s) to connect to")
```
Another instance where my earlier comment applies.

vahidhashemian commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -439,21 +440,30 @@ object ReassignPartitionsCommand extends Logging {
+    val helpOpt = parser.accepts("help", "Print usage information.").forHelp
+
+    var commandDef: String = "This command moves topic partitions between replicas."
+    if(args.length == 0)
```
nit: missing whitespace after `if`

vahidhashemian commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -293,37 +290,48 @@ object TopicCommand extends Logging {
+                         .requiredIf("create", "delete", "alter")
                          .withRequiredArg
-                         .describedAs("topic")
+                         .describedAs("topic to create/alter/describe")
```
And `/delete`?

vahidhashemian commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -335,30 +343,33 @@ object TopicCommand extends Logging {
-    val options = parser.parse(args : _*)
+    val commandDef: String = "Create, delete, describe, or change a topic."
+    
+    if(args.length == 0)
```
nit: missing whitespace after `if`. Seems to be recurring in most of these files.

vahidhashemian commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -335,30 +343,33 @@ object TopicCommand extends Logging {
+
+    val options: OptionSet = CommandLineUtils.tryParse(parser, args)
+    
+    if(options.has("help"))
```
nit: missing whitespace after `if`.

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -255,70 +255,73 @@ object ConsoleConsumer extends Logging {
     val offsetOpt = parser.accepts("offset", "The offset id to consume from (a non-negative number), or 'earliest' which means from beginning, or 'latest' which means from end")
       .withRequiredArg
-      .describedAs("consume offset")
+      .describedAs("consumer offset")
```
`consume offset` is more accurate.

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -64,6 +64,7 @@ object ConsoleProducer {
     } catch {
       case e: joptsimple.OptionException =>
         System.err.println(e.getMessage)
+        e.printStackTrace()
```
Is this required as part of this fix?

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -135,126 +136,135 @@ object ConsoleProducer {
+    val topicOpt = parser.accepts("topic", "The topic id to produce messages to.")
       .withRequiredArg
-      .describedAs("topic")
+      .describedAs("topic id")
```
I think `topic` is better compared to `topic id`, since `topic id` is not often used in the community. Perhaps the above message can also be updated to remove `id`.

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -135,126 +136,135 @@ object ConsoleProducer {
     val batchSizeOpt = parser.accepts("batch-size", "Number of messages to send in a single batch if they are not being sent synchronously.")
       .withRequiredArg
-      .describedAs("size")
+      .describedAs("Number of messages in a single batch")
```
nit: Looks like in most places these descriptions start with a lowercase letter.

vahidhashemian commented on core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala:
```diff
@@ -125,10 +131,8 @@ object ReplicaVerificationTool extends Logging {
-        else
-          false
+        topicMetadata => topicWhiteListFiler.isTopicAllowed(topicMetadata.topic, excludeInternalTopics = false)
+  
```
nit: blank line not needed here.

vahidhashemian commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -269,35 +266,36 @@ object AclCommand {
+      "These are key=val pairs. For the default authorizer, the example values are: zookeeper.connect=localhost:2181.")
       .withRequiredArg
-      .describedAs("authorizer-properties")
+      .describedAs("properties(key=value pairs) for configuring instance of Authorizer")
```
I think we should be careful with the length of the these descriptions because they appear in the help doc of the command. A long description could make the command help difficult to read. It would be great if we can make them more descriptive by using just a few words. If not, we can keep them as is.

vahidhashemian commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -152,38 +152,40 @@ private void parseArguments(final String[] args) throws IOException {
             .withRequiredArg()
             .ofType(String.class)
-            .describedAs("file name");
+            .describedAs("fconfig ile name to be passed to admin clients and embedded consumers");
```
nit: typo in description

vahidhashemian reviewed (commented):
@johnma14 Thanks for addressing my earlier comments. I have a few more (mostly trivial fixes) comments inline.

johnma14 committed:
KAFKA-2111: Add help arguments and required fields

This changeset addresses some minor nits to the description
field for Kafka commands.

From the original commit message describing the overall changes
related to this bug:

This patch addresses the less invasive standardization of command
line arguments as an offshoot of KIP-14. The following changes have
been addressed in this patchset:

- Used the required() method of JOptSimple to identify required options
  for commands. Currently, an option is identified as required by
  mentioning the word REQUIRED in the description of the option.
- Used the requiredIf() and requiredUnless() methods to capture option
  requirements which met the condition - required if/unless some other
  option is present
- Removed calls to CommandLineUtils.checkRequiredArgs(). Since we use
  JOptSimple's required() function to mark options as required, when
  we parse the arguments using OptionParser, it automatically checks
  the required arguments in that call. So there is no need to do this
  check again externally.
- Added a help option to all the commands that were missing it. For
  some of the commands that had the help option, included the call to
  forHelp() method to mark the option as a help option. This means that
  specifying help option on the command line will not cause parsing to
  fail because of missing required options.
- Added try-catch statements to correctly capture OptionException. In
  the case an OptionException is caught, the command usage information
  will be printed along with the exception message. The stacktrace for
  all other exceptions will be printed to System.err. This addresses
  the issue in KAFKA-4220
- Made some minor changes to the description text and added some more
  description for some of the options

johnma14 commented:
retest this please

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -168,23 +167,23 @@ private void parseArguments(final String[] args) throws IOException {
             .withRequiredArg()
             .ofType(String.class)
-            .describedAs("id")
+            .describedAs("application id for Kafka streams")
```
Nit: `[S]treams`

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -168,23 +167,23 @@ private void parseArguments(final String[] args) throws IOException {
             .ofType(String.class)
             .defaultsTo("localhost:9092")
-            .describedAs("urls");
+            .describedAs("server(s) to use for bootstrapping");
```
maybe `server url(s) to ...` ?

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -208,16 +207,18 @@ private void parseArguments(final String[] args) throws IOException {
-        try {
-            options = optionParser.parse(args);
-        } catch (final OptionException e) {
+        if (args.length == 0)
```
Nit: add `{}` to block

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -621,6 +622,7 @@ private void printHelp(OptionParser parser) throws IOException {
                 + "reset tool!\n\n"
         );
         parser.printHelpOn(System.err);
+        Exit.exit(1);
```
We should not exit hard here, but return a appropriate error code if necessary, and propagate it back. Only `main()` should use `exit` to return error code to OS. Using `exit() `makes it hard to test code.

mjsax reviewed (commented):
Only looked into `StreamsResetter` part.

mimaison commented:
This PR is very old and the tools are being rewritten in Java, so I'll close this PR.
